streamer/src/nonblocking/quic.rs

```diff
@@ -960,37 +958,69 @@ async fn packet_batch_sender(
                 Ok(packet_receiver.recv().await)
             };
 
-            if let Ok(Ok(packet_accumulator)) = timeout_res {
+            if let Ok(Ok(mut packet_accumulator)) = timeout_res {
                 // Start the timeout from when the packet batch first becomes non-empty
                 if packet_batch.is_empty() {
                     batch_start_time = Instant::now();
                 }
 
-                unsafe {
-                    let new_len = packet_batch.len() + 1;
-                    packet_batch.set_len(new_len);
-                }
-
-                let i = packet_batch.len() - 1;
-                *packet_batch[i].meta_mut() = packet_accumulator.meta;
+                // 86% of transactions/packets come in one chunk. In that case,
+                // we can just move the chunk to the `Packet` and no copy is
+                // made.
+                // 14% of them come in multiple chunks. In that case, we copy
+                // them into one `Bytes` buffer. We make a copy once, with
+                // intention to not do it again.
+                // We could try to avoid copying even in the second case by
+                // using some kind of chain or rope view over multiple `Bytes`
+                // buffers. The reasons it wasn't done so far are:
```

Comment:
this comment is a bit dramatic. I would cut it here, without explaining the
reasons. Everything in the validator wants a contiguous buffer, it's pretty
obvious, changing that would mean rewriting the validator

Reply:
Done